File: shared/card_rules.lua
--[[
파일명: card_rules.lua
모듈명: CardRules

역할:
- 카드별 수치/제약을 shared/card_rules.json에서 읽어 제공한다.
- mode 기반 필터(SINGLE_ONLY/MULTI_OK)를 지원한다.
- 기존 소비 코드 호환을 위해 tunables를 카드 루트 필드로도 노출한다.

외부에서 사용 가능한 함수:
- CardRules.getRulesVersion()
- CardRules.getCardPool(gameMode)
- CardRules.getCardPoolCount(gameMode)
- CardRules.getCardRule(cardId)
- CardRules.isAllowedInMode(cardId, gameMode)
- CardRules.isTurnCardEnabled(cardId) -- 멀티 기준 호환 함수
- CardRules.getReinforcementRule()
- CardRules.getShockwaveRule()
- CardRules.getInvincibleRule()
- CardRules.getRockfallRule()
- CardRules.getAgileRule()
]]

local Json = require("utils.json")

local CardRules = {}

CardRules.MODE_SINGLE_ONLY = "SINGLE_ONLY"
CardRules.MODE_MULTI_OK = "MULTI_OK"
CardRules.GAME_MODE_SINGLE = "SINGLE"
CardRules.GAME_MODE_MULTI = "MULTI"
CardRules.TARGET_MODE_NONE = "NONE"
CardRules.TARGET_MODE_POINT = "POINT"

local ALLOW_MISSING_MODE_AS_MULTI = false

local DEFAULT_RULES = {
  version = 1,
  card_order = { "reinforcement", "shockwave", "invincible", "rockfall", "agile" },
  cards = {
    reinforcement = {
      id = "reinforcement",
      mode = CardRules.MODE_MULTI_OK,
      enabled = true,
      tags = { "UTILITY", "TRICK" },
      tunables = {
        target_mode = CardRules.TARGET_MODE_POINT,
        min_place_distance = 19,
        lock_spawned_stone_for_turn = true
      }
    },
    shockwave = {
      id = "shockwave",
      mode = CardRules.MODE_MULTI_OK,
      enabled = true,
      tags = { "OFFENSE", "CONTROL" },
      tunables = {
        target_mode = CardRules.TARGET_MODE_NONE,
        radius_multiplier = 4.0,
        strength = 200,
        exclude_source_stone = true,
        ignore_invincible_targets = true
      }
    },
    invincible = {
      id = "invincible",
      mode = CardRules.MODE_MULTI_OK,
      enabled = true,
      tags = { "DEFENSE", "UTILITY" },
      tunables = {
        target_mode = CardRules.TARGET_MODE_NONE,
        protect_after_turn_offset = 1
      }
    },
    rockfall = {
      id = "rockfall",
      mode = CardRules.MODE_MULTI_OK,
      enabled = true,
      tags = { "OBSTACLE", "CONTROL" },
      tunables = {
        target_mode = CardRules.TARGET_MODE_POINT,
        width = 100,
        height = 50,
        margin = 5
      }
    },
    agile = {
      id = "agile",
      mode = CardRules.MODE_MULTI_OK,
      enabled = true,
      tags = { "OFFENSE", "TRICK" },
      tunables = {
        target_mode = CardRules.TARGET_MODE_NONE,
        shot_budget = 2
      }
    }
  }
}

local function deepCopy(value)
  if type(value) ~= "table" then
    return value
  end
  local copied = {}
  for key, nested in pairs(value) do
    copied[key] = deepCopy(nested)
  end
  return copied
end

local function isDevMode()
  if not love or not love.filesystem or type(love.filesystem.isFused) ~= "function" then
    return false
  end
  return love.filesystem.isFused() ~= true
end

local function reportSchemaError(message)
  local normalizedMessage = tostring(message or "unknown_schema_error")
  if isDevMode() then
    error("CardRules schema error: " .. normalizedMessage)
  end
  print("[CardRules] schema warning: " .. normalizedMessage)
end

local function readJsonTable(path)
  if not love or not love.filesystem then
    return nil
  end
  local isReadOk, rawOrError = pcall(love.filesystem.read, path)
  if not isReadOk or type(rawOrError) ~= "string" or rawOrError == "" then
    return nil
  end
  local isDecoded, parsed = pcall(Json.decode, rawOrError)
  if not isDecoded or type(parsed) ~= "table" then
    return nil
  end
  return parsed
end

local function toFiniteNumber(value, fallback)
  if type(value) == "number" and value == value and value ~= math.huge and value ~= -math.huge then
    return value
  end
  return fallback
end

local function toBoolean(value, fallback)
  if type(value) == "boolean" then
    return value
  end
  return fallback
end

local function toTable(value)
  if type(value) == "table" then
    return value
  end
  return {}
end

local function toStringList(value, fallback)
  if type(value) ~= "table" then
    return deepCopy(fallback)
  end

  local seen = {}
  local sanitized = {}
  for _, entry in ipairs(value) do
    if type(entry) == "string" and entry ~= "" and not seen[entry] then
      seen[entry] = true
      sanitized[#sanitized + 1] = entry
    end
  end
  if #sanitized <= 0 then
    return deepCopy(fallback)
  end
  return sanitized
end

local function normalizeMode(rawMode, cardId)
  if rawMode == CardRules.MODE_SINGLE_ONLY or rawMode == CardRules.MODE_MULTI_OK then
    return rawMode
  end

  if ALLOW_MISSING_MODE_AS_MULTI then
    return CardRules.MODE_MULTI_OK
  end

  reportSchemaError("missing_or_invalid_mode:" .. tostring(cardId))
  return CardRules.MODE_MULTI_OK
end

local function normalizeCardRule(cardId, rawCardRule, fallbackCardRule)
  local source = toTable(rawCardRule)
  local fallback = toTable(fallbackCardRule)
  local legacyTunables = {}
  for key, value in pairs(source) do
    if key ~= "id" and key ~= "mode" and key ~= "enabled" and key ~= "tags" and key ~= "tunables" then
      legacyTunables[key] = value
    end
  end

  local fallbackTunables = toTable(fallback.tunables)
  local rawTunables = toTable(source.tunables)
  for key, value in pairs(legacyTunables) do
    rawTunables[key] = value
  end

  local mergedTunables = {}
  for key, value in pairs(fallbackTunables) do
    mergedTunables[key] = value
  end
  for key, value in pairs(rawTunables) do
    mergedTunables[key] = value
  end

  local normalized = {
    id = tostring(source.id or fallback.id or cardId),
    mode = normalizeMode(source.mode or fallback.mode, cardId),
    enabled = toBoolean(source.enabled, toBoolean(fallback.enabled, true)),
    tags = toStringList(source.tags, fallback.tags or {}),
    tunables = mergedTunables
  }

  -- 하위호환: 기존 코드가 cardRule.min_place_distance 같은 루트 필드를 읽어도 동작.
  for key, value in pairs(mergedTunables) do
    normalized[key] = value
  end

  return normalized
end

local function sanitizeRules(raw)
  local source = toTable(raw)
  local fallbackCards = toTable(DEFAULT_RULES.cards)
  local sourceCards = toTable(source.cards)

  local versionValue = source.version
  if versionValue == nil then
    versionValue = source.RULES_VERSION
  end

  local orderSource = source.card_order
  if type(orderSource) ~= "table" then
    orderSource = source.card_pool
  end
  local orderedIds = toStringList(orderSource, DEFAULT_RULES.card_order)
  local seenOrder = {}
  local orderedCardIds = {}
  for _, cardId in ipairs(orderedIds) do
    if not seenOrder[cardId] then
      seenOrder[cardId] = true
      orderedCardIds[#orderedCardIds + 1] = cardId
    end
  end

  for cardId, _ in pairs(sourceCards) do
    local normalizedId = tostring(cardId)
    if normalizedId ~= "" and not seenOrder[normalizedId] then
      seenOrder[normalizedId] = true
      orderedCardIds[#orderedCardIds + 1] = normalizedId
    end
  end

  local normalizedCards = {}
  for _, cardId in ipairs(orderedCardIds) do
    normalizedCards[cardId] = normalizeCardRule(cardId, sourceCards[cardId], fallbackCards[cardId])
  end

  return {
    version = toFiniteNumber(versionValue, DEFAULT_RULES.version),
    card_order = orderedCardIds,
    cards = normalizedCards
  }
end

local loaded = readJsonTable("shared/card_rules.json")
local SANITIZED_RULES = sanitizeRules(loaded)

function CardRules.getRulesVersion()
  return SANITIZED_RULES.version
end

function CardRules.getCardRule(cardId)
  local rule = SANITIZED_RULES.cards[tostring(cardId or "")]
  if type(rule) ~= "table" then
    return nil
  end
  return deepCopy(rule)
end

function CardRules.getCardTunables(cardId)
  local rule = CardRules.getCardRule(cardId)
  if type(rule) ~= "table" then
    return {}
  end
  if type(rule.tunables) ~= "table" then
    return {}
  end
  return deepCopy(rule.tunables)
end

function CardRules.getTargetMode(cardId)
  local tunables = CardRules.getCardTunables(cardId)
  local targetMode = tunables.target_mode
  if targetMode == CardRules.TARGET_MODE_POINT or targetMode == CardRules.TARGET_MODE_NONE then
    return targetMode
  end
  return CardRules.TARGET_MODE_NONE
end

function CardRules.isPointTargetCard(cardId)
  return CardRules.getTargetMode(cardId) == CardRules.TARGET_MODE_POINT
end

function CardRules.isAllowedInMode(cardId, gameMode)
  local rule = SANITIZED_RULES.cards[tostring(cardId or "")]
  if type(rule) ~= "table" or rule.enabled ~= true then
    return false
  end

  if gameMode == CardRules.GAME_MODE_SINGLE then
    return rule.mode == CardRules.MODE_MULTI_OK or rule.mode == CardRules.MODE_SINGLE_ONLY
  end
  return rule.mode == CardRules.MODE_MULTI_OK
end

function CardRules.getCardPool(gameMode)
  local normalizedMode = gameMode or CardRules.GAME_MODE_MULTI
  local copied = {}
  for _, cardId in ipairs(SANITIZED_RULES.card_order or {}) do
    if CardRules.isAllowedInMode(cardId, normalizedMode) then
      copied[#copied + 1] = cardId
    end
  end
  return copied
end

function CardRules.getCardPoolCount(gameMode)
  return #CardRules.getCardPool(gameMode)
end

function CardRules.isTurnCardEnabled(cardId)
  return CardRules.isAllowedInMode(cardId, CardRules.GAME_MODE_MULTI)
end

function CardRules.getReinforcementRule()
  return CardRules.getCardRule("reinforcement")
end

function CardRules.getShockwaveRule()
  return CardRules.getCardRule("shockwave")
end

function CardRules.getInvincibleRule()
  return CardRules.getCardRule("invincible")
end

function CardRules.getRockfallRule()
  return CardRules.getCardRule("rockfall")
end

function CardRules.getAgileRule()
  return CardRules.getCardRule("agile")
end

return CardRules

File: server/src/card_rules.ts
import sharedCardRulesJson from "../../shared/card_rules.json";

export const CARD_MODE_SINGLE_ONLY = "SINGLE_ONLY" as const;
export const CARD_MODE_MULTI_OK = "MULTI_OK" as const;
export type CardMode = typeof CARD_MODE_SINGLE_ONLY | typeof CARD_MODE_MULTI_OK;
export const TARGET_MODE_NONE = "NONE" as const;
export const TARGET_MODE_POINT = "POINT" as const;
export type TargetMode = typeof TARGET_MODE_NONE | typeof TARGET_MODE_POINT;

export const GAME_MODE_SINGLE = "SINGLE" as const;
export const GAME_MODE_MULTI = "MULTI" as const;
export type GameMode = typeof GAME_MODE_SINGLE | typeof GAME_MODE_MULTI;

const ALLOW_MISSING_MODE_AS_MULTI = false;
const STRICT_SCHEMA_IN_DEV = true;

interface BaseCardRule {
  id: string;
  mode: CardMode;
  enabled: boolean;
  tags: string[];
  tunables: Record<string, unknown>;
  [key: string]: unknown;
}

interface ReinforcementCardRule extends BaseCardRule {
  min_place_distance: number;
  lock_spawned_stone_for_turn: boolean;
}

interface ShockwaveCardRule extends BaseCardRule {
  radius_multiplier: number;
  strength: number;
  exclude_source_stone: boolean;
  ignore_invincible_targets: boolean;
}

interface InvincibleCardRule extends BaseCardRule {
  protect_after_turn_offset: number;
}

interface RockfallCardRule extends BaseCardRule {
  width: number;
  height: number;
  margin: number;
}

interface AgileCardRule extends BaseCardRule {
  shot_budget: number;
}

interface CardRules {
  version: number;
  card_order: string[];
  cards: Record<string, BaseCardRule> & {
    reinforcement: ReinforcementCardRule;
    shockwave: ShockwaveCardRule;
    invincible: InvincibleCardRule;
    rockfall: RockfallCardRule;
    agile: AgileCardRule;
  };
}

const DEFAULT_CARD_RULES: CardRules = {
  version: 1,
  card_order: ["reinforcement", "shockwave", "invincible", "rockfall", "agile"],
  cards: {
    reinforcement: {
      id: "reinforcement",
      mode: CARD_MODE_MULTI_OK,
      enabled: true,
      tags: ["UTILITY", "TRICK"],
      tunables: {
        min_place_distance: 19,
        lock_spawned_stone_for_turn: true
      },
      min_place_distance: 19,
      lock_spawned_stone_for_turn: true
    },
    shockwave: {
      id: "shockwave",
      mode: CARD_MODE_MULTI_OK,
      enabled: true,
      tags: ["OFFENSE", "CONTROL"],
      tunables: {
        radius_multiplier: 4,
        strength: 200,
        exclude_source_stone: true,
        ignore_invincible_targets: true
      },
      radius_multiplier: 4,
      strength: 200,
      exclude_source_stone: true,
      ignore_invincible_targets: true
    },
    invincible: {
      id: "invincible",
      mode: CARD_MODE_MULTI_OK,
      enabled: true,
      tags: ["DEFENSE", "UTILITY"],
      tunables: {
        protect_after_turn_offset: 1
      },
      protect_after_turn_offset: 1
    },
    rockfall: {
      id: "rockfall",
      mode: CARD_MODE_MULTI_OK,
      enabled: true,
      tags: ["OBSTACLE", "CONTROL"],
      tunables: {
        width: 100,
        height: 50,
        margin: 5
      },
      width: 100,
      height: 50,
      margin: 5
    },
    agile: {
      id: "agile",
      mode: CARD_MODE_MULTI_OK,
      enabled: true,
      tags: ["OFFENSE", "TRICK"],
      tunables: {
        shot_budget: 2
      },
      shot_budget: 2
    }
  }
};

function isFiniteNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function readNumber(source: Record<string, unknown>, key: string, fallback: number): number {
  const value = source[key];
  return isFiniteNumber(value) ? value : fallback;
}

function readBoolean(source: Record<string, unknown>, key: string, fallback: boolean): boolean {
  const value = source[key];
  return typeof value === "boolean" ? value : fallback;
}

function readObject(source: unknown): Record<string, unknown> {
  if (!source || typeof source !== "object") {
    return {};
  }
  return source as Record<string, unknown>;
}

function readStringListValue(value: unknown, fallback: string[]): string[] {
  if (!Array.isArray(value)) {
    return [...fallback];
  }

  const sanitized: string[] = [];
  const seen = new Set<string>();
  for (const entry of value) {
    if (typeof entry !== "string") {
      continue;
    }
    const normalized = entry.trim();
    if (normalized.length === 0 || seen.has(normalized)) {
      continue;
    }
    seen.add(normalized);
    sanitized.push(normalized);
  }

  if (sanitized.length <= 0) {
    return [...fallback];
  }
  return sanitized;
}

function isDevelopment(): boolean {
  const processRef = (globalThis as { process?: { env?: { NODE_ENV?: string } } }).process;
  return processRef?.env?.NODE_ENV !== "production";
}

function reportSchemaError(message: string): void {
  if (STRICT_SCHEMA_IN_DEV && isDevelopment()) {
    throw new Error(`[card_rules] schema error: ${message}`);
  }
  console.warn(`[card_rules] schema warning: ${message}`);
}

function normalizeMode(rawMode: unknown, cardId: string): CardMode {
  if (rawMode === CARD_MODE_SINGLE_ONLY || rawMode === CARD_MODE_MULTI_OK) {
    return rawMode;
  }
  if (ALLOW_MISSING_MODE_AS_MULTI) {
    return CARD_MODE_MULTI_OK;
  }
  reportSchemaError(`missing_or_invalid_mode:${cardId}`);
  return CARD_MODE_MULTI_OK;
}

function normalizeCardRule(
  cardId: string,
  rawCardRule: Record<string, unknown>,
  fallbackCardRule: BaseCardRule
): BaseCardRule {
  const legacyTunables: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(rawCardRule)) {
    if (key === "id" || key === "mode" || key === "enabled" || key === "tags" || key === "tunables") {
      continue;
    }
    legacyTunables[key] = value;
  }

  const rawTunables = {
    ...readObject(fallbackCardRule.tunables),
    ...readObject(rawCardRule.tunables),
    ...legacyTunables
  };

  const normalized: BaseCardRule = {
    id: typeof rawCardRule.id === "string" && rawCardRule.id.length > 0 ? rawCardRule.id : fallbackCardRule.id || cardId,
    mode: normalizeMode(rawCardRule.mode ?? fallbackCardRule.mode, cardId),
    enabled: readBoolean(rawCardRule, "enabled", fallbackCardRule.enabled === true),
    tags: readStringListValue(rawCardRule.tags, fallbackCardRule.tags || []),
    tunables: rawTunables
  };

  // 하위호환: 기존 abilities.ts가 rule.shot_budget 같은 루트 필드를 그대로 참조할 수 있게 유지.
  for (const [key, value] of Object.entries(rawTunables)) {
    normalized[key] = value;
  }

  return normalized;
}

function sanitizeCardRules(raw: unknown): CardRules {
  const source = readObject(raw);
  const cardsSource = readObject(source.cards);
  const fallbackCards = DEFAULT_CARD_RULES.cards;
  const versionValue = source.version ?? source.RULES_VERSION;
  const orderSource = Array.isArray(source.card_order) ? source.card_order : source.card_pool;

  const seen = new Set<string>();
  const orderedCardIds: string[] = [];
  for (const cardId of readStringListValue(orderSource, DEFAULT_CARD_RULES.card_order)) {
    if (!seen.has(cardId)) {
      seen.add(cardId);
      orderedCardIds.push(cardId);
    }
  }
  for (const cardId of Object.keys(cardsSource)) {
    if (!seen.has(cardId)) {
      seen.add(cardId);
      orderedCardIds.push(cardId);
    }
  }

  const normalizedCards: Record<string, BaseCardRule> = {};
  for (const cardId of orderedCardIds) {
    const rawCardRule = readObject(cardsSource[cardId]);
    const fallbackCardRule = fallbackCards[cardId] || {
      id: cardId,
      mode: CARD_MODE_MULTI_OK,
      enabled: false,
      tags: [],
      tunables: {}
    };
    normalizedCards[cardId] = normalizeCardRule(cardId, rawCardRule, fallbackCardRule);
  }

  return {
    version: isFiniteNumber(versionValue) ? versionValue : DEFAULT_CARD_RULES.version,
    card_order: orderedCardIds,
    cards: normalizedCards as CardRules["cards"]
  };
}

export const CARD_RULES = sanitizeCardRules(sharedCardRulesJson);

export function getCardRule(cardId: string): BaseCardRule | null {
  const rule = CARD_RULES.cards[cardId];
  return rule ? { ...rule } : null;
}

export function isAllowedInMode(cardId: string, gameMode: GameMode): boolean {
  const rule = CARD_RULES.cards[cardId];
  if (!rule || rule.enabled !== true) {
    return false;
  }
  if (gameMode === GAME_MODE_SINGLE) {
    return rule.mode === CARD_MODE_MULTI_OK || rule.mode === CARD_MODE_SINGLE_ONLY;
  }
  return rule.mode === CARD_MODE_MULTI_OK;
}

export function getTargetMode(cardId: string): TargetMode {
  const rule = CARD_RULES.cards[cardId];
  if (!rule || typeof rule.tunables !== "object" || !rule.tunables) {
    return TARGET_MODE_NONE;
  }
  const rawMode = (rule.tunables as Record<string, unknown>).target_mode;
  if (rawMode === TARGET_MODE_POINT || rawMode === TARGET_MODE_NONE) {
    return rawMode;
  }
  return TARGET_MODE_NONE;
}

export function isPointTargetCard(cardId: string): boolean {
  return getTargetMode(cardId) === TARGET_MODE_POINT;
}

function buildCardPoolByMode(gameMode: GameMode): string[] {
  const pool: string[] = [];
  for (const cardId of CARD_RULES.card_order) {
    if (isAllowedInMode(cardId, gameMode)) {
      pool.push(cardId);
    }
  }
  return pool;
}

export const CARD_POOL = buildCardPoolByMode(GAME_MODE_MULTI);
export const CARD_POOL_SET = new Set<string>(CARD_POOL);
export const SINGLE_CARD_POOL = buildCardPoolByMode(GAME_MODE_SINGLE);

export function isTurnCardEnabled(cardId: string): boolean {
  return isAllowedInMode(cardId, GAME_MODE_MULTI);
}
